cpp_linter/common_fs/__init__.py

```diff
@@ -158,6 +159,86 @@ def is_range_contained(self, start: int, end: int) -> Optional[Tuple[int, int]]:
+            Defaults to 1 second.
+
+        :returns: The bytes read from the file.
+
+        :raises FileIOTimeout: When the operation did not succeed due to a timeout.
+        :raises OSError: When the file could not be opened due to an `OSError`.
+        """
+        contents = b""
+        success = False
+        exception: Union[OSError, FileIOTimeout] = FileIOTimeout(
+            "Failed to read from file within %d seconds"
+            % round(timeout_ns / 1_000_000_000, 2)
+        )
+        timeout = time.monotonic_ns() + timeout_ns
+        while not success and time.monotonic_ns() < timeout:
+            try:
+                with open(self.name, "rb") as f:
+                    while not success and time.monotonic_ns() < timeout:
+                        if f.readable():
+                            contents = f.read()
+                            success = True
+            except OSError as exc:  # pragma: no cover
+                exception = exc
+        if not success and exception:  # pragma: no cover
+            raise exception
+        return contents
+
+    def read_write_with_timeout(
+        self,
+        data: Union[bytes, bytearray],
+        timeout_ns: int = 1_000_000_000,
+    ) -> bytes:
+        """Read then write the entire file's contents.
+
+        :param data: The bytes to write to the file. This will overwrite the contents
+            being read beforehand.
+        :param timeout_ns: The number of nanoseconds to wait till timeout occurs.
+            Defaults to 1 second.
+
+        :returns: The bytes read from the file.
+
+        :raises FileIOTimeout: When the operation did not succeed due to a timeout.
+        :raises OSError: When the file could not be opened due to an `OSError`.
+        """
+        success = False
+        exception: Union[OSError, FileIOTimeout] = FileIOTimeout(
+            "Failed to read then write to file within %d seconds"
+            % round(timeout_ns / 1_000_000_000, 2)
+        )
+        original_data = b""
+        timeout = time.monotonic_ns() + timeout_ns
+        while not success and time.monotonic_ns() < timeout:
+            try:
+                with open(self.name, "r+b") as f:
+                    while not success and time.monotonic_ns() < timeout:
+                        if f.readable():
+                            original_data = f.read()
+                            f.seek(0)
+                        else:  # pragma: no cover
+                            continue
+                        while not success and time.monotonic_ns() < timeout:
+                            if f.writable():
+                                f.write(data)
+                                success = True
+            except OSError as exc:  # pragma: no cover
+                exception = exc
+        if not success and exception:  # pragma: no cover
+            raise exception
+        return original_data
+
```

Comment:
I thought about this, but it would only incur more runtime overhead because the algorithms are not as simple as swapping out a function call. And we'd be unsafely passing around an open file handle.